target_chains/starknet/contracts/src/reader.cairo

```diff
@@ -0,0 +1,333 @@
+use core::option::OptionTrait;
+use core::array::ArrayTrait;
+use core::keccak::cairo_keccak;
+use core::integer::u128_byte_reverse;
+use core::fmt::{Debug, Formatter};
+
+pub const EOF: felt252 = 'unexpected end of input';
+pub const UNEXPECTED_OVERFLOW: felt252 = 'unexpected overflow';
+
+/// A byte array with storage format similar to `core::ByteArray`, but
+/// suitable for reading data from it.
+#[derive(Drop, Clone, Serde)]
+pub struct ByteArray {
+    // Number of bytes stored in the last item of `self.data` (or 0 if it's empty).
+    num_last_bytes: u8,
+    // Bytes in big endian. Each item except the last one stores 31 bytes.
+    // If `num_last_bytes < 31`, unused most significant bytes of the last item will be unused.
+    data: Array<bytes31>,
+}
+
+impl DebugByteArray of Debug<ByteArray> {
+    fn fmt(self: @ByteArray, ref f: Formatter) -> Result<(), core::fmt::Error> {
+        write!(f, "ByteArray {{ num_last_bytes: {}, data: [", self.num_last_bytes)?;
+        let mut data = self.data.clone();
+        loop {
+            match data.pop_front() {
+                Option::Some(v) => {
+                    let v: u256 = v.into();
+                    write!(f, "{:?}, ", v).unwrap();
+                },
+                Option::None => { break; },
+            }
+        };
+        write!(f, "]}}")
+    }
+}
+
+#[generate_trait]
+pub impl ByteArrayImpl of ByteArrayTrait {
+    /// Creates a byte array with the data.
+    fn new(data: Array<bytes31>, num_last_bytes: u8) -> ByteArray {
+        if data.len() == 0 {
+            assert!(num_last_bytes == 0);
+        } else {
+            assert!(num_last_bytes <= 31);
+        // TODO: check that unused bytes are zeroed.
+        }
+        ByteArray { num_last_bytes, data }
+    }
+
+    /// Removes 31 or less bytes from the start of the array.
+    /// Returns the value and the number of bytes.
+    fn pop_front(ref self: ByteArray) -> Option<(bytes31, u8)> {
+        let item = self.data.pop_front()?;
+        if self.data.is_empty() {
+            let num_bytes = self.num_last_bytes;
+            self.num_last_bytes = 0;
+            Option::Some((item, num_bytes))
+        } else {
+            Option::Some((item, 31))
+        }
+    }
+
+    fn len(self: @ByteArray) -> usize {
+        if self.data.is_empty() {
+            0
+        } else {
+            (self.data.len() - 1) * 31 + (*self.num_last_bytes).into()
+        }
+    }
+}
+
+/// Allows to read data from a byte array.
+/// Uses big endian unless specified otherwise.
+/// All methods return `EOF` error if attempted to
+/// read more bytes than is available.
+#[derive(Drop, Clone)]
+pub struct Reader {
```

Comment:
I don't think it's practical to avoid Drop in cairo. It's really hard to get anything working without Drop because all arithmetic operations can diverge, so you can't even calculate anything when there is a non-Drop type in scope. The same goes for error propagation and unwraps.